File: src/lib/agents/llm-provider.ts
/**
 * Text-chat LLM call for the agent layer. Reuses GOOGLE_API_KEY and the same
 * generativelanguage REST endpoint as providers/gemini.ts and providers/omni.ts
 * (direct fetch, no SDK — matches every other provider in this codebase)
 * rather than adding a new provider/credential for chat.
 *
 * Model defaults to Google's "gemini-flash-latest" alias, which tracks
 * whatever Google currently serves under that name — avoids hardcoding a
 * specific dated model id this session can't verify. Override with
 * AGENT_LLM_MODEL. Verify the contract with scripts/probe-agent-llm.ts before
 * relying on this in production, same convention as this repo's other probes.
 */
import type { AgentMessage } from "./types";

const API_ROOT = "https://generativelanguage.googleapis.com/v1beta";
const DEFAULT_MODEL = "gemini-flash-latest";

export function agentModel(): string {
  return process.env.AGENT_LLM_MODEL || DEFAULT_MODEL;
}

export interface CallLLMInput {
  model?: string;
  systemPrompt: string;
  messages: AgentMessage[];
}

export interface CallLLMResult {
  text: string;
  usage?: { tokensIn?: number; tokensOut?: number };
}

export async function callLLM({
  model,
  systemPrompt,
  messages,
}: CallLLMInput): Promise<CallLLMResult> {
  const apiKey = process.env.GOOGLE_API_KEY;
  if (!apiKey) throw new Error("GOOGLE_API_KEY is not set.");

  // Gemini's generateContent has no "system"/"assistant" role — system
  // instructions are a separate field, and its assistant role is "model".
  const contents = messages
    .filter((m) => m.role !== "system")
    .map((m) => ({
      role: m.role === "assistant" ? "model" : "user",
      parts: [{ text: m.content }],
    }));

  const body = {
    systemInstruction: { parts: [{ text: systemPrompt }] },
    contents,
  };

  const res = await fetch(`${API_ROOT}/models/${model || agentModel()}:generateContent`, {
    method: "POST",
    headers: { "Content-Type": "application/json", "x-goog-api-key": apiKey },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    const errText = await res.text();
    throw new Error(`Agent LLM error (${res.status}): ${errText.slice(0, 400)}`);
  }

  const json = await res.json();
  const part = (json?.candidates?.[0]?.content?.parts ?? []).find(
    (p: { text?: string }) => typeof p?.text === "string"
  );
  if (!part) {
    const reason = json?.candidates?.[0]?.finishReason || "no candidates";
    throw new Error(`Agent LLM returned no text (${reason}).`);
  }

  const usageMeta = json?.usageMetadata;
  return {
    text: part.text,
    usage: usageMeta
      ? {
          tokensIn: usageMeta.promptTokenCount,
          tokensOut: usageMeta.candidatesTokenCount,
        }
      : undefined,
  };
}

File: backend/apps/agents/agents/llm_provider.py
"""Port of src/lib/agents/llm-provider.js — text-chat LLM call for the
agent layer. Reuses GOOGLE_API_KEY and the same generativelanguage REST
endpoint as providers/gemini.py and providers/omni.py."""

import os

import requests

API_ROOT = "https://generativelanguage.googleapis.com/v1beta"
DEFAULT_MODEL = "gemini-flash-latest"


def agent_model() -> str:
    return os.environ.get("AGENT_LLM_MODEL", DEFAULT_MODEL)


def call_gemini_raw(system_prompt: str, contents: list[dict], tools: list[dict] | None = None, model: str | None = None) -> dict:
    """contents: [{"role": "user"|"model", "parts": [GeminiPart, ...]}, ...].
    Returns {"parts", "finishReason", "usage"}."""
    api_key = os.environ.get("GOOGLE_API_KEY")
    if not api_key:
        raise RuntimeError("GOOGLE_API_KEY is not set.")

    body: dict = {"systemInstruction": {"parts": [{"text": system_prompt}]}, "contents": contents}
    if tools:
        body["tools"] = [{"functionDeclarations": tools}]

    res = requests.post(
        f"{API_ROOT}/models/{model or agent_model()}:generateContent",
        headers={"Content-Type": "application/json", "x-goog-api-key": api_key},
        json=body,
        timeout=60,
    )
    if not res.ok:
        raise RuntimeError(f"Agent LLM error ({res.status_code}): {res.text[:400]}")

    data = res.json()
    candidate = (data.get("candidates") or [{}])[0]
    usage_meta = data.get("usageMetadata")
    return {
        "parts": (candidate.get("content") or {}).get("parts") or [],
        "finishReason": candidate.get("finishReason"),
        "usage": (
            {"tokensIn": usage_meta.get("promptTokenCount"), "tokensOut": usage_meta.get("candidatesTokenCount")}
            if usage_meta else None
        ),
    }


def call_llm(system_prompt: str, messages: list[dict], model: str | None = None) -> dict:
    """messages: [{"role": "user"|"assistant"|"system", "content": str}, ...].
    Returns {"text", "usage"}."""
    contents = [
        {"role": "model" if m["role"] == "assistant" else "user", "parts": [{"text": m["content"]}]}
        for m in messages if m["role"] != "system"
    ]
    result = call_gemini_raw(system_prompt, contents, model=model)
    part = next((p for p in result["parts"] if isinstance(p.get("text"), str)), None)
    if not part:
        raise RuntimeError(f"Agent LLM returned no text ({result.get('finishReason') or 'no candidates'}).")
    return {"text": part["text"], "usage": result.get("usage")}
